File: JAVA/129 - Sum Root to Leaf Numbers.java
/*
 129. Sum Root to Leaf Numbers (M)

 Approach 1 : DFS
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
    public int sumNumbers(TreeNode root) {
        return DFS(root, 0);
    }
    
    public int DFS(TreeNode root, int prev) {
        if (root == null) return 0;
        int sum = prev * 10 + root.val;
        if (root.left == null && root.right == null) return sum;
        return DFS(root.left, sum) + DFS(root.right, sum);
    }
}

File: C++/129-Sum_Root_to_Leaf_Numbers.cpp
/*
 129. Sum Root to Leaf Numbers (M)

 A1: DFS
 DFS the tree and save the node value, if the node is a leaf, add it to the total sum.
 Time: O(n)
 Space: O(n)

*/

// A1
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int sum = 0;
        DFS(root, sum, 0);
        return sum;
    }
    
    void DFS(TreeNode* root, int& sum, int curSum) {
        if (!root) return;
        curSum = curSum * 10 + root->val;

        // If the node is a leaf, add the current sum to the total sum
        if (!root->left && !root->right) sum += curSum;

        // If not a leaf, move into it
        if (root->left) DFS(root->left, sum, curSum);
        if (root->right) DFS(root->right, sum, curSum);
    }
};
